worldedit-core/src/main/java/com/sk89q/worldedit/math/convolution/SnowHeightMap.java

```diff
@@ -0,0 +1,213 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.math.convolution;
+
+import com.sk89q.worldedit.EditSession;
+import com.sk89q.worldedit.MaxChangedBlocksException;
+import com.sk89q.worldedit.function.mask.Mask;
+import com.sk89q.worldedit.math.BlockVector3;
+import com.sk89q.worldedit.regions.Region;
+import com.sk89q.worldedit.registry.state.Property;
+import com.sk89q.worldedit.world.block.BlockState;
+import com.sk89q.worldedit.world.block.BlockTypes;
+
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Allows applications of Kernels onto the region's height map for snow layers
+ *
+ * <p>Currently only used for snow layer smoothing (with a GaussianKernel)</p>.
+ */
+public class SnowHeightMap {
+    private final float[] data;
+    private final int width;
+    private final int height;
+
+    private final Region region;
+    private final EditSession session;
+
+    private final Property<Object> layers;
+
+    /**
+     * Constructs the SnowHeightMap.
+     *
+     * @param session an edit session
+     * @param region the region
+     * @param mask optional mask for the height map
+     */
+    public SnowHeightMap(EditSession session, Region region, @Nullable Mask mask) {
+        checkNotNull(session);
+        checkNotNull(region);
+
+        this.session = session;
+        this.region = region;
+
+        this.width = region.getWidth();
+        this.height = region.getLength();
+
+        layers = BlockTypes.SNOW.getProperty("layers");
+
+        int minX = region.getMinimumPoint().getBlockX();
+        int minY = region.getMinimumPoint().getBlockY();
+        int minZ = region.getMinimumPoint().getBlockZ();
+        int maxY = region.getMaximumPoint().getBlockY();
+
+        // Store current heightmap data
+        data = new float[width * height];
+        for (int z = 0; z < height; ++z) {
+            for (int x = 0; x < width; ++x) {
+                int highestBlockY = session.getHighestTerrainBlock(x + minX, z + minZ, minY, maxY, mask);
+                BlockState upper = session.getBlock(BlockVector3.at(x + minX, highestBlockY + 1, z + minZ));
+                if (upper.getBlockType() == BlockTypes.SNOW) {
+                    Integer amountLayers = (Integer) upper.getState(layers);
+                    data[z * width + x] = (highestBlockY + 1 + (((float) amountLayers - 1) / 8));
+                } else {
+                    BlockState block = session.getBlock(BlockVector3.at(x + minX, highestBlockY, z + minZ));
+                    if (block.getBlockType() == BlockTypes.AIR) {
```

Comment:
This should be using the `block.getBlockType().getMaterial().isAir()` method rather than doing this equality check